nixos/modules/services/databases/victoriametrics/default.nix

```diff
@@ -0,0 +1,186 @@
+      map (filterAttrsListRecursive pred) x
+    else
+      x;
+in
+{
+  options.services.victoriametrics = {
+    enable = mkEnableOption "VictoriaMetrics, a time series database, long-term remote storage for victoriametrics";
+    package = mkPackageOption pkgs "victoriametrics" { };
+
+    port = mkOption {
+      type = types.port;
+      default = 8428;
+      description = ''
+        Port to listen on.
+      '';
+    };
+
+    listenAddress = mkOption {
+      type = types.str;
+      default = "0.0.0.0";
+      description = ''
+        The listen address for the http interface.
+      '';
+    };
+
+    stateDir = mkOption {
+      type = types.str;
+      default = "victoriametrics";
+      description = ''
+        Directory below `/var/lib` to store VictoriaMetrics metrics data.
+        This directory will be created automatically using systemd's StateDirectory mechanism.
+      '';
+    };
+
+    retentionPeriod = mkOption {
+      type = types.nullOr types.str;
+      default = null;
+      example = "15d";
+      description = ''
+        How long to retain samples in storage.
+        The minimum retentionPeriod is 24h or 1d. See also -retentionFilter
+        The following optional suffixes are supported: s (second), h (hour), d (day), w (week), y (year).
+        If suffix isn't set, then the duration is counted in months (default 1)
+      '';
+    };
+
+    globalConfig = mkOption {
+      type = promTypes.globalConfig;
+      default = { };
+      description = ''
+        Parameters that are valid in all  configuration contexts. They
+        also serve as defaults for other configuration sections
+      '';
+    };
+
+    scrapeConfigs = mkOption {
+      type = types.listOf promTypes.scrape_config;
+      default = [ ];
+      description = ''
+        A list of scrape configurations.
+        See docs: <https://docs.victoriametrics.com/vmagent/#how-to-collect-metrics-in-prometheus-format>
+      '';
+    };
+
+    extraOptions = mkOption {
+      type = types.listOf types.str;
+      default = [ ];
+      description = ''
+        Extra options to pass to VictoriaMetrics. See the README:
+        <https://github.com/VictoriaMetrics/VictoriaMetrics/blob/master/README.md>
+        or {command}`victoriametrics -help` for more
+        information.
+      '';
+    };
+  };
+  config = lib.mkIf cfg.enable {
+    systemd.services.victoriametrics = {
+      description = "VictoriaMetrics time series database";
+      wantedBy = [ "multi-user.target" ];
+      after = [ "network.target" ];
+      startLimitBurst = 5;
+
+      serviceConfig = {
+        ExecStart = ''
+          ${cfg.package}/bin/victoria-metrics \
+              -storageDataPath=${workingDir} \
+              -httpListenAddr=${bindAddr} \
+              -retentionPeriod=${cfg.retentionTime} \
+              -promscrape.config=${generatedPrometheusYml} \
+              ${lib.escapeShellArgs cfg.extraFlags}
+        '';
+        DynamicUser = true;
```

Comment:
Prometheus uses a hard-coded user, defined in https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/misc/ids.nix . Can I ask we do the same here?

Reply:
(I'd add that moving to a static/set user becomes important if `vmagent` needs secrets to scrape sources of metrics- passing an agenix encrypted secret to a service using `DynamicUser` isn't feasible, I don't think?)

Reply:
Why not? That works fine if it's passed via systemd tools (Load credentials).
The vmagent module is capable of doing so.

https://github.com/NixOS/nixpkgs/blob/18536bf04cd71abd345f9579158841376fdd0c5a/nixos/modules/services/monitoring/vmagent.nix#L95